Sources/SwiftGenKit/Parsers/Files/FilesParser.swift

```diff
@@ -0,0 +1,47 @@
+//
+// SwiftGenKit
+// Copyright © 2020 SwiftGen
+// MIT Licence
+//
+
+import CoreServices
+import Foundation
+import PathKit
+
+public enum Files {
+  public enum ParserError: Error, CustomStringConvertible {
+    case invalidFile(path: Path, reason: String)
+
+    public var description: String {
+      switch self {
+      case .invalidFile(let path, let reason):
+        return "Unable to parse file at \(path). \(reason)"
+      }
+    }
+  }
+
+  // MARK: Resource File Parser
+
+  public class Parser: SwiftGenKit.Parser {
+    private let options: ParserOptionValues
+    var files: [File] = []
+    public var warningHandler: Parser.MessageHandler?
+
+    public required init(options: [String: Any] = [:], warningHandler: Parser.MessageHandler? = nil) throws {
+      self.options = try ParserOptionValues(options: options, available: Parser.allOptions)
+      self.warningHandler = warningHandler
+    }
+
+    public static var defaultFilter: String = ".*"
+
+    public func parse(path: Path, relativeTo parent: Path) throws {
+      if path.isDirectory {
```

Comment:
This check is not necessary: when you get here, you have a path to a file that matches your regex.
The actual recursion happens here:
https://github.com/SwiftGen/SwiftGen/blob/stable/Sources/SwiftGenKit/Parsers/Parser.swift#L57-L70

Reply:
Are you saying all I should need is `files.append(try File(path: path, relativeTo: parent))`? That doesn't work with the default filter of `.*`. The path passed in to `searchAndParse` then matches the folder and the recursion never happens

Reply:
Unrelated to how we implement recursion, I'm wondering if we should make either the filter or a parser option allow the user to opt-in or out of recursion?

E.g. when using glob patterns in shell (or ruby) you can use `foo/**/*.txt` to mean "every txt file inside the foo folder however deep in the hierarchy it is" while `foo/*.txt` will only match txt files that are directly under the foo folder. (`man fnmatch` for more info)

Obviously we can't use that here because we use `NSRegularExpressions` for our filters, not glob/fnmatch. So that won't be a solution for us. But opens the question is providing a way to have similar choice of behavior would be useful (maybe via a parser option)? wdyt?

Reply:
@mgray88 Hmmm right, that's an issue here because we don't have extensions to match on. We could change the default to `.*\..*`, to ensure we match files with extensions 🤷 (not the best option, I know)

Another option may be to add a flag to the parser protocol to indicate whether a parser only wants files or not. Could also be an option set, so a parser could say `wantFiles | wantDirectories | wantHiddenFiles`. (Ignore the actual option names, just an example). If needed, I can create a separate PR for this, and we can rebase this PR on that.

@AliSoftware not sure when users would want to disable recursion? 99% of users won't ever have to, and that 1% that has loads of files and deep directory structures, can avoid any performance impact by moving the right resources into a subfolder and point SwiftGen at that.

Reply:
@djbe I wouldn't know where to start with that 😆. I'm sure you could probably implement that much quicker

Related also is the test below not using `searchAndParse`

Reply:
Just finished adding a `relativeTo` option to the Files parser, which I think might be a better option than going in and modifying the parser protocol (unless you see uses for that elsewhere)